packages/opencensus-web-instrumentation-zone/src/xhr-interceptor.ts

```diff
@@ -0,0 +1,162 @@
+ * instance of `XMLHttpRequest`.
+ * In case the task is intercepted, sets the Trace Context Header to it and
+ * creates a child span related to this XHR in case it is OPENED.
+ * In case the XHR is DONE, end the child span.
+ */
+export function interceptXhrTask(task: AsyncTask) {
+  if (!isTrackedTask(task)) return;
+  if (!(task.target instanceof XMLHttpRequest)) return;
+
+  const xhr = task.target as XhrWithUrl;
+  if (xhr.readyState === XMLHttpRequest.OPENED) {
+    incrementXhrTaskCount();
+    const rootSpan: RootSpan = task.zone.get('data').rootSpan;
+    setTraceparentContextHeader(xhr, rootSpan);
+  } else if (xhr.readyState === XMLHttpRequest.DONE) {
+    endXhrSpan(xhr);
+    decrementXhrTaskCount();
+  }
+}
+
+function setTraceparentContextHeader(
+  xhr: XhrWithUrl,
+  rootSpan: RootSpan
+): void {
+  // `__zone_symbol__xhrURL` is set by the Zone monkey-path.
+  const xhrUrl = xhr.__zone_symbol__xhrURL;
+  const childSpan = rootSpan.startChildSpan({
+    name: parseUrl(xhrUrl).pathname,
+    kind: SpanKind.CLIENT,
+  });
+  // Associate the child span to the XHR so it allows to
+  // find the correct span when the request is DONE.
+  xhrSpans.set(xhr, childSpan);
+  if (traceOriginMatchesOrSameOrigin(xhrUrl)) {
+    xhr.setRequestHeader(
+      TRACEPARENT_HEADER,
+      spanContextToTraceParent({
+        traceId: rootSpan.traceId,
+        spanId: childSpan.id,
+      })
+    );
+  }
+}
+
+function endXhrSpan(xhr: XhrWithUrl): void {
+  const span = xhrSpans.get(xhr);
+  if (span) {
+    // TODO: Investigate more to send the the status code a `number` rather
+    // than `string`. Once it is able to send as a number, change it.
+    span.addAttribute(ATTRIBUTE_HTTP_STATUS_CODE, xhr.status.toString());
+    span.addAttribute(ATTRIBUTE_HTTP_METHOD, xhr._ocweb_method);
+    span.end();
+    joinPerfResourceDataToSpan(xhr, span);
+    xhrSpans.delete(xhr);
+  }
+}
+
+/**
+ * If xhr task count is 0, clear the Performance Resource Timings.
+ * This is done in order to help the browser Performance resource timings
+ * selector algorithm to take only the data related to the current XHRs running.
+ */
+function maybeClearPerfResourceBuffer(): void {
+  if (xhrTasksCount === 0) {
+    performance.clearResourceTimings();
+    // Clear the set as these entries are not longer necessary.
+    alreadyAssignedPerfEntries.clear();
+  }
+}
+
+function joinPerfResourceDataToSpan(xhr: XhrWithUrl, span: Span) {
+  const xhrPerfResourceTiming = getXhrPerfomanceData(xhr.responseURL, span);
+  if (!xhrPerfResourceTiming) return;
+
+  alreadyAssignedPerfEntries.add(xhrPerfResourceTiming.mainRequest);
+  if (xhrPerfResourceTiming.corsPreFlightRequest) {
+    alreadyAssignedPerfEntries.add(xhrPerfResourceTiming.corsPreFlightRequest);
+    const corsPerfTiming = xhrPerfResourceTiming.corsPreFlightRequest as PerformanceResourceTimingExtended;
+    setCorsPerfTimingAsChildSpan(corsPerfTiming, span);
```

Comment:
When there's a cors pre-flight request, would it make sense to add the main request as a child span too? (I'm not saying you definitely should, just curious what you think)

Reply:
I think that might be another option too, however, as the span is already related to the XHR I think we just want to see some performance annotation related to the main request and also, as we are sending the `traceparent` context header to the server, we will be able to see the span related to that specific request. otherwise, if the `traceparent` is not sent, the annotations will be enough I think.